jax/_src/scipy/optimize/bfgs.py

```diff
@@ -133,7 +136,15 @@ def body_fun(state):
         failed=line_search_results.failed,
         line_search_status=line_search_results.status,
     )
-    s_k = line_search_results.a_k * p_k
+    alpha_k = line_search_results.a_k
+    # If the step is too small in 32 bit mode then the optimizer can get stuck
+    # with a direction of zero - ensure that doesn't happen by setting a
+    # minimum step size.
+    if not jax.config.x64_enabled:
+      alpha_k = jnp.where(jnp.abs(alpha_k) < 1e-8,
+                          jnp.sign(alpha_k) * 1e-8,
```

Comment:
These thresholds looks suspiciously close to zero for float32. As a general rule of thumb, floating point arithmetic is accuracy up to some multiple of `np.finfo(dtype).eps`, which is roughly 1e-7 for float32 and 2e-16 for float64:
https://en.wikipedia.org/wiki/Machine_epsilon